File: test/e2e/Apps/BasicDotNetIsolated/TerminateOrchestration.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Net;
using Grpc.Core;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Client;
using Microsoft.Extensions.Logging;

namespace Microsoft.Azure.Durable.Tests.E2E
{
    public static class LongRunningOrchestration
    {
        [Function(nameof(LongRunningOrchestrator))]
        public static async Task<List<string>> LongRunningOrchestrator(
            [OrchestrationTrigger] TaskOrchestrationContext context)
        {
            ILogger logger = context.CreateReplaySafeLogger(nameof(HelloCities));
            logger.LogInformation("Starting long-running orchestration.");
            var outputs = new List<string>();

            // Call our fake activity 100,000 times to simulate an orchestration that might run for >= 10,000s (2.7 hours)
            for (int i = 0; i < 100000; i++) 
            {
                outputs.Add(await context.CallActivityAsync<string>(nameof(SimulatedWorkActivity), 100));
            }

            return outputs;
        }

        [Function(nameof(SimulatedWorkActivity))]
        public static string SimulatedWorkActivity([ActivityTrigger]int sleepMs, FunctionContext executionContext)
        {
            // Sleep the provided number of ms to simulate a long-running activity operation
            ILogger logger = executionContext.GetLogger("SimulatedWorkActivity");
            logger.LogInformation("Sleeping for {sleepMs}ms.", sleepMs);
            Thread.Sleep(sleepMs);
            return $"Slept for {sleepMs}ms.";
        }

        [Function("LongOrchestrator_HttpStart")]
        public static async Task<HttpResponseData> HttpStart(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
            [DurableClient] DurableTaskClient client,
            FunctionContext executionContext)
        {
            ILogger logger = executionContext.GetLogger("LongOrchestrator_HttpStart");

            string instanceId = await client.ScheduleNewOrchestrationInstanceAsync(
                nameof(LongRunningOrchestrator));

            logger.LogInformation("Started orchestration with ID = '{instanceId}'.", instanceId);
            
            return await client.CreateCheckStatusResponseAsync(req, instanceId);
        }

        [Function("TerminateInstance")]
        public static async Task<HttpResponseData> Run(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
            [DurableClient] DurableTaskClient client,
            string instanceId)
        {
            string reason = "Long-running orchestration was terminated early.";
            try 
            {
                await client.TerminateInstanceAsync(instanceId, reason);
                return req.CreateResponse(HttpStatusCode.OK);
            }
            catch (RpcException ex) 
            {
                var response = req.CreateResponse(HttpStatusCode.BadRequest);
                response.Headers.Add("Content-Type", "text/plain");
                await response.WriteStringAsync(ex.Message);
                return response;
            }
        }
    }
}

File: test/e2e/Apps/BasicJava/src/main/java/com/function/TerminateOrchestration.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.microsoft.azure.functions.*;

import java.util.*;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableActivityTrigger;
import com.microsoft.durabletask.azurefunctions.DurableClientContext;
import com.microsoft.durabletask.azurefunctions.DurableClientInput;
import com.microsoft.durabletask.azurefunctions.DurableOrchestrationTrigger;

import java.util.logging.Logger;


/**
 * Orchestration and activity functions for simulating long-running orchestration and termination.
 */
public class TerminateOrchestration {

    /**
     * This is the orchestrator function.
     */
    @FunctionName("LongRunningOrchestrator")
    public List<String> longRunningOrchestrator(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx,
            final ExecutionContext context) {
        Logger logger = context.getLogger();
        logger.info("Starting long-running orchestration.");
        List<String> outputs = new ArrayList<>();

        // Call our fake activity 100,000 times to simulate an orchestration that might run for >= 10,000s (2.7 hours)
        for (int i = 0; i < 100000; i++) {
            outputs.add(ctx.callActivity("SimulatedWorkActivity", 100, String.class).await());
        }

        return outputs;
    }

    /**
     * This is the activity function.
     */
    @FunctionName("SimulatedWorkActivity")
    public String simulatedWorkActivity(
            @DurableActivityTrigger(name = "sleepMs") int sleepMs,
            final ExecutionContext context) {
        context.getLogger().info("Sleeping for " + sleepMs + "ms.");
        try {
            Thread.sleep(sleepMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return "Sleep interrupted.";
        }
        return "Slept for " + sleepMs + "ms.";
    }

    /**
     * HTTP-triggered function to terminate an orchestration instance.
     */
    @FunctionName("TerminateInstance")
    public HttpResponseMessage run(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            @BindingName("instanceId") String instanceId,
            final ExecutionContext context) {
        DurableTaskClient client = durableContext.getClient();
        String reason = "Long-running orchestration was terminated early.";
        try {
            client.terminate(instanceId, reason);
            return request.createResponseBuilder(HttpStatus.OK).build();
        } catch (Exception ex) {
            context.getLogger().severe(ex.getMessage());
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                    .header("Content-Type", "text/plain")
                    .body(ex.getMessage())
                    .build();
        }
    }
}